File: src/MyServiceBus.RabbitMq/RabbitMqReceiveEndpointTopology.cs
using MyServiceBus.Topology;

namespace MyServiceBus;

public sealed record RabbitMqReceiveEndpointTopology(
    string QueueName,
    string ExchangeName,
    string RoutingKey,
    string ExchangeType,
    bool Durable,
    bool AutoDelete,
    ushort PrefetchCount,
    IDictionary<string, object?>? QueueArguments)
{
    public static RabbitMqReceiveEndpointTopology Project(ReceiveEndpointTopology endpoint)
    {
        ArgumentNullException.ThrowIfNull(endpoint);
        ArgumentException.ThrowIfNullOrWhiteSpace(endpoint.QueueName);
        ArgumentException.ThrowIfNullOrWhiteSpace(endpoint.ExchangeName);
        ArgumentException.ThrowIfNullOrWhiteSpace(endpoint.ExchangeType);

        if (endpoint.Durable && endpoint.AutoDelete)
            throw new ArgumentException("A RabbitMQ receive endpoint cannot be both durable and auto-delete.", nameof(endpoint));

        return new RabbitMqReceiveEndpointTopology(
            endpoint.QueueName,
            endpoint.ExchangeName,
            endpoint.RoutingKey,
            endpoint.ExchangeType,
            endpoint.Durable,
            endpoint.AutoDelete,
            endpoint.PrefetchCount,
            endpoint.QueueArguments);
    }
}

File: src/Java/myservicebus-rabbitmq/src/main/java/com/myservicebus/rabbitmq/RabbitMqReceiveEndpointTopology.java
package com.myservicebus.rabbitmq;

import com.myservicebus.topology.MessageBinding;
import com.myservicebus.topology.ReceiveEndpointTransportTopology;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public record RabbitMqReceiveEndpointTopology(
        String queueName,
        List<MessageBinding> bindings,
        int prefetchCount,
        Map<String, Object> queueArguments) {

    public RabbitMqReceiveEndpointTopology {
        if (queueName == null || queueName.isBlank()) {
            throw new IllegalArgumentException("RabbitMQ receive endpoint queue name cannot be blank");
        }
        if (bindings == null || bindings.isEmpty()) {
            throw new IllegalArgumentException("RabbitMQ receive endpoint must have at least one binding");
        }
        if (bindings.stream().anyMatch(binding -> binding.getEntityName() == null || binding.getEntityName().isBlank())) {
            throw new IllegalArgumentException("RabbitMQ receive endpoint binding entity name cannot be blank");
        }
        if (prefetchCount < 0) {
            throw new IllegalArgumentException("RabbitMQ receive endpoint prefetch count cannot be negative");
        }

        bindings = List.copyOf(bindings);
        queueArguments = queueArguments == null
                ? null
                : Collections.unmodifiableMap(new LinkedHashMap<>(queueArguments));
    }

    public static RabbitMqReceiveEndpointTopology project(
            String queueName,
            List<MessageBinding> bindings,
            int prefetchCount,
            Map<String, Object> queueArguments) {
        return new RabbitMqReceiveEndpointTopology(queueName, bindings, prefetchCount, queueArguments);
    }

    public static RabbitMqReceiveEndpointTopology project(ReceiveEndpointTransportTopology endpoint) {
        if (endpoint.durable() && endpoint.temporary()) {
            throw new IllegalArgumentException("A RabbitMQ receive endpoint cannot be both durable and temporary");
        }
        return new RabbitMqReceiveEndpointTopology(
                endpoint.name(),
                endpoint.bindings(),
                endpoint.prefetchCount(),
                endpoint.transportOptions());
    }
}
